kernel/src/log_compaction/writer.rs

```diff
@@ -0,0 +1,265 @@
+            let missing: Vec<Version> = expected_versions
+                .iter()
+                .filter(|v| !found_versions.contains(v))
+                .copied()
+                .collect();
+
+            return Err(Error::generic(format!(
+                "Expected {} commit files for range [{}, {}], but found {}. Missing versions: {:?}",
+                expected_count,
+                self.start_version,
+                self.end_version,
+                commit_files.len(),
+                missing
+            )));
+        }
+
+        // Create a log segment from these commit files
+        let log_segment = self.create_log_segment(commit_files)?;
+
+        // Read actions from the log segment - the segment handles reading commits in the right order
+        let actions_iter = log_segment.read_actions(
+            engine,
+            COMPACTION_ACTIONS_SCHEMA.clone(),
+            COMPACTION_ACTIONS_SCHEMA.clone(),
+            None, // No predicate - we want all actions
+        )?;
+
+        // Create checkpoint log replay processor for compaction
+        // This reuses the same reconciliation logic as checkpoints
+        // Use None for txn_expiration_timestamp as we typically want to preserve all transactions in compaction
+        let processor = CheckpointLogReplayProcessor::new(
+            self.min_file_retention_timestamp_millis,
+            None, // txn_expiration_timestamp
+        );
+
+        // Process actions using the same iterator pattern as checkpoints
+        // The processor handles reverse chronological processing internally
+        let result_iter = processor.process_actions_iter(actions_iter);
+
+        // Collect processed batches into a vector for the iterator
+        let mut batches = Vec::new();
+        let mut total_actions = 0i64;
+        let mut total_add_actions = 0i64;
+
+        for batch_result in result_iter {
+            let batch = batch_result?;
+            // Track statistics
+            total_actions += batch.actions_count;
+            total_add_actions += batch.add_actions_count;
+            // Extract the data from the batch - pass the FilteredEngineData directly
+            batches.push(batch.filtered_data.data);
+        }
+
+        Ok(LogCompactionDataIterator {
+            batches: batches.into_iter(),
+            total_actions,
+            total_add_actions,
+        })
+    }
+
+    /// Finalize the compaction after the data has been written
+    ///
+    /// This method completes the compaction process and can perform any
+    /// necessary cleanup or validation.
+    #[allow(dead_code)]
+    pub(crate) fn finalize(
+        self,
+        _engine: &dyn Engine,
+        _metadata: &FileMeta,
+        _data_iterator: LogCompactionDataIterator,
+    ) -> DeltaResult<()> {
+        // For now, just validate that the iterator was exhausted
+        // In the future, we might want to validate file metadata, etc.
+        Ok(())
+    }
```

Comment:
Do we need `finalize` in compaction? For checkpoints we rely on it to write out the `_last_checkpoint` file but I don't think we really have anything similar for compaction?